packages/contracts/contracts/adapters/UniswapV3Adapter.sol

```diff
@@ -0,0 +1,122 @@
+    ISwapRouter public immutable _swapRouter;
+
+    // Whitelist of tokens that can be used in a path
+    mapping(address => bool) private _allowedTokens;
+
+    /// @notice Event emitted when a token is added to the whitelist
+    event TokenPermissionSet(address token, bool permission);
+
+    // Constructor, sets uniswap swap router contract
+    constructor(address swapRouter) Ownable() {
+        _swapRouter = ISwapRouter(swapRouter);
+    }
+
+    /// @notice Set token permission in whitelist
+    /// @param token Token address
+    /// @param permission Whether token is allowed
+    function setTokenPermission(
+        address token,
+        bool permission
+    ) external onlyOwner {
+        _allowedTokens[token] = permission;
+        emit TokenPermissionSet(token, permission);
+    }
+
+    /// @notice Proxy for exactInputSingle function of Uniswap SwapRouter. Ensures tokenIn
+    ///         is allowed and recipient is caller.
+    /// @param params ExactInputSingleParams (see ISwapRouter)
+    function exactInputSingle(
+        ISwapRouter.ExactInputSingleParams memory params
+    ) external returns (uint256 amountOut) {
+        require(_allowedTokens[params.tokenIn], "!allowed token");
+        require(params.recipient == msg.sender, "!recipient");
+
+        IERC20(params.tokenIn).transferFrom(
+            address(msg.sender),
+            address(this),
+            params.amountIn
+        );
+        IERC20(params.tokenIn).approve(address(_swapRouter), params.amountIn);
+
+        amountOut = _swapRouter.exactInputSingle(params);
+    }
+
+    /// @notice Proxy for exactInput function of Uniswap SwapRouter.
+    ///         Ensures all tokens in path are is allowed and recipient is caller.
+    /// @param params ExactInputParams (see ISwapRouter)
+    function exactInput(
+        ISwapRouter.ExactInputParams memory params
+    ) external returns (uint256 amountOut) {
+        require(tokensAreAllowed(params.path), "!allowed path");
+        require(params.recipient == msg.sender, "!recipient");
+
+        address tokenIn = extractTokenAddressFromPath(params.path, 0);
+
+        IERC20(tokenIn).transferFrom(
+            address(msg.sender),
+            address(this),
+            params.amountIn
+        );
+        IERC20(tokenIn).approve(address(_swapRouter), params.amountIn);
+
+        amountOut = _swapRouter.exactInput(params);
+    }
+
+    /// @notice Checks that all tokens in path are allowed, returns false if any are not
+    /// @param path Path of tokens
+    /// @dev path is formatted as [address token1][uint24 fee][address token2], thus the length will
+    ///      always be ((20 + 3) * n) + 20 bytes where n is the number of hops
+    function tokensAreAllowed(bytes memory path) internal view returns (bool) {
+        uint256 length = path.length;
+
+        // Ensure the path length is correct: ((20 + 3) * n) + 20
+        require((length - 20) % 23 == 0, "Invalid path length");
```

Comment:
re: segregating all of the pointer arithmetic, can we extract this into a separate internal function as well?